src/common/latest/prepare-task.ts

```diff
@@ -65,9 +72,40 @@ export const prepareTask: TaskJob = async (endpointType: EndpointType) => {
 
   tl.setVariable(TaskVariables.SonarQubeScannerParams, jsonParams);
 
+  // always download the scanner, uses default versions if not explicitly overriden
+  const scannerPath = await downloadScanner(cliVersion, msBuildVersion);
+
+  tl.setVariable(TaskVariables.SonarQubeScannerLocation, scannerPath);
+
+  // we need to pass in the downloaded paths here
   await scanner.runPrepare();
 };
 
+async function downloadScanner(cliVersion?: string, msBuildVersion?: string) {
+  const scannerMode = tl.getVariable(TaskVariables.SonarQubeScannerMode);
+  const msBuildFileUrl = isWindows() ? scannerConfig.classicUrl : scannerConfig.dotnetUrl;
+
+  let fileUrl = scannerMode === ScannerMode.CLI ? scannerConfig.cliUrl : msBuildFileUrl;
+
+  // Override the default URL if a version is specified
+  if (cliVersion) {
+    fileUrl = scannerConfig.cliUrlTemplate(cliVersion);
+  } else if (msBuildVersion) {
+    fileUrl = scannerConfig.msBuildUrlTemplate(msBuildVersion, isWindows());
+  }
+
+  tl.debug(`Downloading scanner from ${fileUrl}`);
+  const downloadPath = await toolLib.downloadTool(fileUrl);
+  tl.debug(`Downloaded: ${fileUrl} file to ${downloadPath}`);
+
+  tl.debug(`Extracting ${downloadPath}`);
+  const unzipPath = await toolLib.extractZip(downloadPath);
+  tl.debug(`Unzipped file to ${unzipPath}`);
+  // `downloadPath` now contains the path to the downloaded file
+
+  return unzipPath;
+}
```

Comment:
Potentially, using try/catch here would allow us to make the error more understandable to the user when the fetching fails, right now that's what the user sees when he specifies a version that does not exist
![image](https://github.com/SonarSource/sonar-scanner-vsts/assets/31401273/06a49933-e71e-4747-9976-df6ccc8dc7df)